kernel.c

```diff
@@ -0,0 +1,728 @@
+        // Mark it as used
+        frame_free[i] |= (1 << j);
+        return i * BITS_PER_BYTE + j;
+      }
+    }
+  }
+  return -1;
+}
+
+/**
+ * Mark a frame as used in the bit-vector
+ */
+void get_frame_number(int index) {
+  frame_free[index / 8] |= (1 << (index % 8));
+}
+
+/**
+ * Free a previously allocated frame in the bit-vector
+ */
+void free_frame_number(int index) {
+  frame_free[index / 8] &= ~(1 << (index % 8));
+}
+
+/**
+ * Set up the page tables for kernel and initial user context
+ */
+static void SetupPageTable(void) {
+  // Allocate and zero kernel page table
+  kernel_page_table = calloc(MAX_PT_LEN, sizeof(pte_t));
+  if (kernel_page_table == NULL) Halt();
+
+  // Invalidate all entries initially
+  for (int i = 0; i < MAX_PT_LEN; i++) {
+    kernel_page_table[i].valid = 0;
+  }
+
+  // Map text pages: read+exec, one-to-one mapping
+  for (int p = _first_kernel_text_page; p < _first_kernel_data_page; p++) {
+    kernel_page_table[p].valid = 1;
+    kernel_page_table[p].prot = PROT_READ | PROT_EXEC;
+    get_frame_number(p);            // Reserve this frame
+    kernel_page_table[p].pfn = p;   // PFN equals page number
+  }
+
+  // Map data pages: read+write
+  for (int p = _first_kernel_data_page; p < _orig_kernel_brk_page; p++) {
+    kernel_page_table[p].valid = 1;
+    kernel_page_table[p].prot = PROT_READ | PROT_WRITE;
+    get_frame_number(p);
+    kernel_page_table[p].pfn = p;
+  }
+
+  // Map kernel stack region
+  for (int p = KERNEL_STACK_BASE >> PAGESHIFT;
+       p < KERNEL_STACK_LIMIT >> PAGESHIFT;
+       p++) {
+    kernel_page_table[p].valid = 1;
+    kernel_page_table[p].pfn   = p;
+    get_frame_number(p);
+    kernel_page_table[p].prot  = PROT_READ | PROT_WRITE;
+  }
+
+  // Allocate empty user page table
+  user_page_table = calloc(MAX_PT_LEN, sizeof(pte_t));
+  if (user_page_table == NULL) Halt();
+  for (int i = 0; i < MAX_PT_LEN; i++) {
+    user_page_table[i].valid = 0;
+  }
+
+  // Give the user stack one free frame at top of space
+  int index = MAX_PT_LEN - 1;
+  int frame = get_free_frame();
+  user_page_table[index].valid = 1;
+  user_page_table[index].pfn   = frame;
+  user_page_table[index].prot  = PROT_READ | PROT_WRITE;
+
+  // Load base/limit registers for kernel (PTBR0) and user (PTBR1)
+  WriteRegister(REG_PTBR0, (unsigned int)kernel_page_table);
+  WriteRegister(REG_PTLR0, MAX_PT_LEN);
+  WriteRegister(REG_PTBR1, (unsigned int)user_page_table);
+  WriteRegister(REG_PTLR1, MAX_PT_LEN);
+}
+
+/**
+ * Kernel entry point: initializes memory, traps, and idle process
+ */
+void KernelStart(char *cmd_args[], unsigned int pmem_size,
+                 UserContext *uctxt) {
+  // Save original break point
+  vm_on_brk = _orig_kernel_brk_page;
+
+  // Compute total physical pages and initialize free map
+  num_phys_pages = (int)(pmem_size >> PAGESHIFT) / 8;
+  frame_free = calloc(num_phys_pages, sizeof(unsigned char));
+  for (int i = 0; i < num_phys_pages; i++) {
```

Comment:
nit: there's a few times throughout the codebase where you manually set things to zero right after a calloc call, which is unnecessary. not a big deal